### ISSUE TITLE
There is no variable named "var"

### ISSUE THREAD

**Issue by danilomnds** [hashicorp/terraform#35679]
### Terraform Version

```shell
Terraform v1.9.5
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v4.0.1
```

### Terraform Configuration Files

I'm working on a terraform module for azure virtual machine scale set. The block "network_interface" is required and can be used multiple times. The same applies for the nested block "ip_configuration".

It seems that nested blocks are not getting the "var.network_interface" content. 

I wasn't able to find a solution for this... do you have any ideia?

You can check all the testing details here https://github.com/danilomnds/terraform-azurerm-vmss-uniform

```terraform
resource "azurerm_linux_virtual_machine_scale_set" "vmss_uniform" {
  name                = var.name
  resource_group_name = var.resource_group_name
  location            = var.location
  admin_username      = var.admin_username
  instances           = var.instances
  sku                 = var.sku
  # first network interface
  network_interface {
    name = var.network_interface[0].name
    # first ip configuration
    ip_configuration {
      name                                         = var.network_interface[0].ip_configuration[0].name
      application_gateway_backend_address_pool_ids = lookup(var.network_interface[0].ip_configuration[0], "application_gateway_backend_address_pool_ids", null)
      application_security_group_ids               = lookup(var.network_interface[0].ip_configuration[0], "application_security_group_ids", null)
      load_balancer_backend_address_pool_ids       = lookup(var.network_interface[0].ip_configuration[0], "load_balancer_backend_address_pool_ids", null)
      load_balancer_inbound_nat_rules_ids          = lookup(var.network_interface[0].ip_configuration[0], "load_balancer_inbound_nat_rules_ids", null)
      primary                                      = lookup(var.network_interface[0].ip_configuration[0], "primary", null)      
      dynamic "public_ip_address" {
        for_each = var.network_interface[0].ip_configuration[0].public_ip_address != null ? [var.network_interface[0].ip_configuration[0].public_ip_address] : []
        content {
          name                    = public_ip_address.value.name
          domain_name_label       = lookup(public_ip_address.value, "domain_name_label", null)
          idle_timeout_in_minutes = lookup(public_ip_address.value, "idle_timeout_in_minutes", null)
          dynamic "ip_tag" {
            for_each = public_ip_address.value.ip_tag != null ? [public_ip_address.value.ip_tag] : []
            content {
              tag  = ip_tag.value.tag
              type = ip_tag.value.type
            }
          }
          public_ip_prefix_id = lookup(public_ip_address.value, "public_ip_prefix_id", null)
          version             = lookup(public_ip_address.value, "version", null)
        }
      }
      subnet_id = lookup(var.network_interface[0].ip_configuration[0], "subnet_id", null)
      version   = lookup(var.network_interface[0].ip_configuration[0], "version", null)
    }
```
 

### Debug Output

n/a

### Expected Behavior

I'd like to load the content of var.network_interface[0].ip_configuration[0].public_ip_address without getting the error "There is no variable named "var""

### Actual Behavior

error message

 Error: Unknown variable
│
│   on .terraform\modules\vmss-prd-devops-002\main.tf line 20, in resource "azurerm_linux_virtual_machine_scale_set" "vmss_uniform":
│   20:         for_each = var.network_interface[0].ip_configuration[0].public_ip_address != null ? [var.network_interface[0].ip_configuration[0].public_ip_address] : []
│
│ There is no variable named "var".
╵
╷
│ Error: Unknown variable
│
│   on .terraform\modules\vmss-prd-devops-002\main.tf line 41, in resource "azurerm_linux_virtual_machine_scale_set" "vmss_uniform":
│   41:       for_each = length(var.network_interface[0].ip_configuration) > 1 ? slice(var.network_interface[0].ip_configuration, 1, length(var.network_interface[0].ip_configuration)) : []
│
│ There is no variable named "var".
╵
╷
│ Error: Unknown variable
│
│   on .terraform\modules\vmss-prd-devops-002\main.tf line 78, in resource "azurerm_linux_virtual_machine_scale_set" "vmss_uniform":
│   78:     for_each = length(var.network_interface) > 1 ? slice(var.network_interface, 1, length(var.network_interface)) : []
│
│ There is no variable named "var".
╵
......
│   on .terraform\modules\vmss-prd-devops-002\main.tf line 148, in resource "azurerm_linux_virtual_machine_scale_set" "vmss_uniform":
│  148:     for_each = var.automatic_os_upgrade_policy != null ? [var.automatic_os_upgrade_policy] : []
│
│ There is no variable named "var".
╵

### Steps to Reproduce

1) terraform plan

**Comment by liamcervante** [comment]
Hi @danilomnds, thanks for filing this. The issue you are facing is coming from line 202 in your main.tf: https://github.com/danilomnds/terraform-azurerm-vmss-uniform/blob/main/main.tf#L202.

```
for_each = data_disk.value.protected_settings_from_key_vault != null ? [data_disk.value.protected_settings_from_key_vault] : []
```

You have defined `data_disk` as a dynamic block on line 171, but that is not available within the scope of the dynamic block defined at line 201, which is only within scope of the `extension` dynamic block.

The errors about the `var` variable are red herrings that are the result of Terraform failing to build a successful evaluation context as a result of the real error, so it has not variables available to it at all when it tries to evaluate the other dynamic blocks within the configuration. Fixing that "true" error should hopefully unblock you.

Ideally, though, Terraform wouldn't throw up these fake errors to confuse you so I think we can keep this as a bug and look into making sure only the proper error is shared.

**Comment by liamcervante** [comment]
Duplicate of https://github.com/hashicorp/terraform/issues/34188
